src/createNativeStackNavigator.tsx

```diff
@@ -0,0 +1,465 @@
+import React from 'react';
+import {
+  NativeSyntheticEvent,
+  NativeTouchEvent,
+  Platform,
+  StyleSheet,
+  Animated,
+  StyleProp,
+  TextStyle,
+  ViewStyle,
+} from 'react-native';
+import { ScreenStackHeaderConfigProps, StackPresentationTypes } from './types';
+import {
+  Screen,
+  ScreenStack,
+  ScreenStackHeaderBackButtonImage,
+  ScreenStackHeaderCenterView,
+  ScreenStackHeaderConfig,
+  ScreenStackHeaderLeftView,
+  ScreenStackHeaderRightView,
+} from './index.native';
+import {
+  createNavigator,
+  SceneView,
+  StackActions,
+  StackRouter,
+  NavigationRouteConfigMap,
+  CreateNavigatorConfig,
+  NavigationStackRouterConfig,
+  NavigationParams,
+  NavigationRoute,
+  NavigationDescriptor,
+  NavigationState,
+  NavigationNavigator,
+  NavigationAction,
+} from 'react-navigation';
+import { NativeStackNavigationOptions } from './native-stack/types';
+import { HeaderBackButton } from 'react-navigation-stack';
+import {
+  StackNavigationHelpers,
+  StackNavigationProp,
+  Layout,
+} from 'react-navigation-stack/src/vendor/types';
+
+const REMOVE_ACTION = 'NativeStackNavigator/REMOVE';
+
+function renderComponentOrThunk(componentOrThunk: unknown, props: unknown) {
+  if (typeof componentOrThunk === 'function') {
+    return componentOrThunk(props);
+  }
+  return componentOrThunk;
+}
+
+type NativeStackRemoveNavigationAction = {
+  type: 'NativeStackNavigator/REMOVE';
+  immediate: boolean;
+  key?: string;
+};
+
+type ExtendedNativeStackNavigationOptions = StackNavigatorOptions &
+  NativeStackNavigationOptions &
+  BackButtonProps & {
+    onWillAppear: () => void;
+    onAppear: () => void;
+    onWillDisappear: () => void;
+    onDisappear: () => void;
+    // these props differ from the ones used in v5 `native-stack`, and we would like to keep the API consistent between versions
+    /** @deprecated Use headerHideShadow */
+    hideShadow: boolean;
+    /** @deprecated Use headerLargeTitle */
+    largeTitle: boolean;
+    /** @deprecated Use headerLargeTitleHideShadow */
+    largeTitleHideShadow: boolean;
+    /** @deprecated Use headerTranslucent */
+    translucent: boolean;
+  };
+
+// these are adopted from `stack` navigator, and probably should be changed to the ones used by v5 `native-stack`
```

Comment:
TODO: think about how to treat the types that should be deprecated (READMEs and code).